Add bookings

svlgerasimov opened the pull request.

svlgerasimov committed:
refactor: заменить хранилища User и Item на JPA репозитории

svlgerasimov committed:
feat: добавить таблицу и entity для бронирования

svlgerasimov committed:
refactor: маппинг UserDto через Mapstruct

svlgerasimov committed:
refactor: маппинг ItemDto через Mapstruct

svlgerasimov committed:
feat: добавить бронирования

svlgerasimov committed:
feat: добавить комментарии к предметам

svlgerasimov committed:
style: убрать лишнее

svlgerasimov committed:
style: пустые строки перед методами

TyutterinYakov commented on pom.xml:
```diff
@@ -40,10 +46,16 @@
 			<artifactId>spring-boot-configuration-processor</artifactId>
 			<optional>true</optional>
 		</dependency>
+<!--		<dependency>-->
```
Если код не используется - следует его исключить

TyutterinYakov commented on pom.xml:
```diff
@@ -24,10 +24,16 @@
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-data-jpa</artifactId>
+<!--			<version>3.0.0</version>-->
```
Если код не используется - следует его исключить

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,68 @@
+    @PostMapping
+    public BookingDtoOut add(@RequestBody @Valid BookingDtoIn bookingDto,
+                             @RequestHeader("X-Sharer-User-Id") long userId) {
+        if (bookingDto.getEnd().isBefore(bookingDto.getStart())) {
```
Не забываем проверить, что время старта не равно времени окончания)

(предложение)Эту валидацию можно было бы реализовать с помощью аннотации валидации над классом, чтобы поддерживать декларативный подход https://devcolibri.com/spring-mvc-кастомная-аннотация-для-валидации/
Кстати, если будет интересно то, как реализовать кастомную аннотацию валидации на уровне класса `BookingDtoIn` - можешь чекнуть тут) https://github.com/TyutterinYakov/CustomValidationClassLevel Я написал небольшой примерчик, чтобы было проще разобраться в теме)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,68 @@
+import javax.validation.Valid;
+import java.util.List;
+
+/**
```
Комментарий более не несет смысла, следует его исключить здесь и в других классах, где уже реализован функционал)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,68 @@
+    public List<BookingDtoOut> findByBooker(
+            @RequestParam(required = false, defaultValue = "ALL") String state,
+            @RequestHeader("X-Sharer-User-Id") long userId) {
+        try {
```
Почему бы эту логику не вынести в отдельный приватный метод, чтобы не дублировать код) Либо же вынести в Enum state в виде статического метода?)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,68 @@
+    public List<BookingDtoOut> findByOwner(
+            @RequestParam(required = false, defaultValue = "ALL") String state,
+            @RequestHeader("X-Sharer-User-Id") long userId) {
+        try {
```
Почему бы эту логику не вынести в отдельный приватный метод, чтобы не дублировать код) Либо же вынести в Enum state в виде статического метода?)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,68 @@
+
+    @GetMapping
+    public List<BookingDtoOut> findByBooker(
+            @RequestParam(required = false, defaultValue = "ALL") String state,
```
Если указано `defaultValue`, это уже подразумевает, что параметр не является обязательным, следовательно `required = false` не требуется)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/controller/BookingController.java:
```diff
@@ -0,0 +1,68 @@
+
+    @GetMapping("/owner")
+    public List<BookingDtoOut> findByOwner(
+            @RequestParam(required = false, defaultValue = "ALL") String state,
```
Если указано `defaultValue`, это уже подразумевает, что параметр не является обязательным, следовательно `required = false` не требуется)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDtoIn.java:
```diff
@@ -0,0 +1,22 @@
+
+    @NotNull
+    @Future
+    LocalDateTime start;
```
Время старта бронирования может быть вполне равно текущей дате) Хорошо было бы использовать более подходящую аннотацию

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/model/Booking.java:
```diff
@@ -1,20 +1,52 @@
+@Table(name = "bookings")
+@Getter
+@Setter
+@ToString
```
Перед использованием аннотаций генерации кода в классах-сущностях, хорошо было бы посмотреть на проблемы, которые могут возникнуть
https://habr.com/ru/company/haulmont/blog/564682/

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/model/Comment.java:
```diff
@@ -0,0 +1,46 @@
+@Table(name = "comments")
+@Getter
+@Setter
+@ToString
```
Перед использованием аннотаций генерации кода в классах-сущностях, хорошо было бы посмотреть на проблемы, которые могут возникнуть
https://habr.com/ru/company/haulmont/blog/564682/

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/model/Item.java:
```diff
@@ -1,15 +1,43 @@
+@Table(name = "items")
+@Getter
+@Setter
+@ToString
```
Перед использованием аннотаций генерации кода в классах-сущностях, хорошо было бы посмотреть на проблемы, которые могут возникнуть
https://habr.com/ru/company/haulmont/blog/564682/

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/model/User.java:
```diff
@@ -1,10 +1,36 @@
+@Table(name = "users")
+@Getter
+@Setter
+@ToString
```
Перед использованием аннотаций генерации кода в классах-сущностях, хорошо было бы посмотреть на проблемы, которые могут возникнуть
https://habr.com/ru/company/haulmont/blog/564682/

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -19,58 +19,60 @@
 @RequiredArgsConstructor(onConstructor = @__(@Autowired))
 @Slf4j
 public class UserServiceImpl implements UserService {
```
Так как ты познакомился с транзакциями, хорошо было бы все сервисы сделать транзакционными)
Для этого необходимо навесить над ними аннотацию 
`@Transactional(readOnly = true)`, что указывает на то, что фиксация изменений не будет производиться) А над теми методами, где требуется фиксация в БД, необходимо использовать `@Transactional`

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -19,58 +19,60 @@
         if (Objects.nonNull(patchEmail)) {
             // если пришёл патч запрос с той же почтой, что и была раньше - ничего страшного
-            userStorage.getByEmailExcludeId(patchEmail, id).ifPresent(
+            userRepository.findByEmailAndIdIsNot(patchEmail, id).ifPresent(
```
Действительно ли требуется эта проверка? На уровне БД у тебя введен индекс для поддержания уникальности почты и при попытке сохранить запись с дублирующей почтой, возникнет `DataIntegrityViolationException`, которое можно поймать в хендлере и обработать с требуемым статусом)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/storage/BookingRepository.java:
```diff
@@ -0,0 +1,54 @@
+
+    List<Booking> findByBookerAndEndIsBeforeOrderByStartDesc(User booker, LocalDateTime maxEnd);
+
+    List<Booking> findByBookerAndStatusIsOrderByStartDesc(User booker, BookingStatus status);
```
Можно было бы оставить возможность выбора сортировки, чтобы в случае, когда нам потребуется другой порядок - не писать еще один метод) Для этого в аргументах метода следует так же ожидать Sort sort.
Объект этого класса создается подобным образом: Sort.by(Sort.Direction.DESC, "start");
А OrderByStartDesc следовательно нужно будет убрать)
Аналогично с другими методами репозитория, где может потребоваться возможность выбора сортировки

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -4,74 +4,124 @@
+                formDtoExtended(item);
+    }
+
+    private ItemDtoOutExtended formDtoExtended(Item item) {
```
Хорошо было бы придерживаться порядка во всех классах. Сначала располагать публичные методы - предоставляемое API, а уже потом приватные - внутренняя реализация

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -4,74 +4,124 @@
-                .map(ItemMapper::toItemDto)
+    public List<ItemDtoOutExtended> getAll(long userId) {
+        User owner = getUser(userId);
+        return itemRepository.findAllByOwner(owner).stream()
```
Следует получить комментарии всех вещей, которые будут возвращены вне цикла и разбить их на мапу, где ключ - идентификатор вещи, а значение - коллекция из комментариев) Таким образом мы не будем производить n-лишних запросов к БД) Получение следует произвести вне цикла)
Аналогично следует поступить и с поиском предыдущего и следующего бронирования)
Где можно было бы по коллекции предметов произвести получение и сразу же выполнить сортировку, чтобы поиск необходимых бронирований был проще(получать бронирования следует только в состоянии APPROVED)
```
		Map<Item, List<Comment>> comments = commentRepository.findByItemIn(items, Sort.by(DESC, "created"))
				.stream()
				.collect(groupingBy(Comment::getItem, toList()));
```
Это лишь предложение по реализации, можно реализовать свою логику, главное избавить об обращения к репозиториям в циклах)

TyutterinYakov reviewed (commented):
Добрый вечер! Отметил то, что следует исправить и то, что можно сделать лучше) Хорошего дня! Если возникнут вопросы - пиши в пачку, буду рад помочь!

svlgerasimov committed:
style: удалить закомментированный код из pom.xml

svlgerasimov committed:
refactor: заменить проверку времени бронирования на кастомную аннотацию
fix: заменить в dto бронирования аннотацию над start на FutureOrPresent
fix: исправить обработчик исключения MethodArgumentNotValidException, чтобы корректно отображать ответ для ошибки валидации объекта целиком

svlgerasimov committed:
style: убрать TODO ветки add-bookings

svlgerasimov committed:
refactor: вынести метод с преобразованием и проверкой корректности BookingSearchState в сам enum

svlgerasimov committed:
refactor: убрать из RequestParam required там, где есть defaultValue

svlgerasimov committed:
fix: в сущностях убрать из toString "ленивые" поля

svlgerasimov committed:
feat: добавить Transactional в сервисы

svlgerasimov committed:
refactor: добавить в ErrorHandler обработку DataIntegrityViolationException

svlgerasimov committed:
refactor: в запросных методах перенести тип сортировки в аргумент

svlgerasimov committed:
style: порядок методов - сначала реализация интерфейса, потом приватные

svlgerasimov committed:
refactor: убрать n+1 из ItemServiceImpl

svlgerasimov committed:
fix: следующее и последнее бронирование искать только со статусом APPROVED

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDtoIn.java:
```diff
@@ -0,0 +1,25 @@
+@BookingStartBeforeEnd
+public class BookingDtoIn {
+
+    @NotNull
```
Аннотация `@NotNull` не требуется, так как проверка на null входит в кастомную валидацию)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/booking/dto/BookingDtoIn.java:
```diff
@@ -0,0 +1,25 @@
+    @FutureOrPresent
+    LocalDateTime start;
+
+    @NotNull
```
Аннотация `@NotNull` не требуется, так как проверка на null входит в кастомную валидацию)

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -3,75 +3,150 @@
+                .collect(Collectors.groupingBy(Comment::getItem, Collectors.toList()));
+
+        LocalDateTime now = LocalDateTime.now();
+        List<Booking> lastBookings = bookingRepository.findAllByItemInAndStartBeforeAndStatusIs(
```
Время старта вполне может быть равно текущему времени, то есть последнее бронирование в это же время может выступать как текущее

TyutterinYakov commented on src/main/java/ru/practicum/shareit/item/service/ItemServiceImpl.java:
```diff
@@ -3,75 +3,150 @@
+
+    private ItemDtoOutExtended formDtoExtendedWithBookings(Item item) {
+        LocalDateTime now = LocalDateTime.now();
+        Booking lastBooking = bookingRepository
```
Время старта вполне может быть равно текущему времени, то есть последнее бронирование в это же время может выступать как текущее

TyutterinYakov commented on src/main/java/ru/practicum/shareit/user/service/UserServiceImpl.java:
```diff
@@ -3,74 +3,71 @@
-        }
-        UserMapper.patchWithUserDto(user, patchDto);
+        userPatchDtoMapper.updateWithPatchDto(user, patchDto);
+        user = userRepository.save(user);
```
Вызов метода save более не будет требоваться, так как мы будем работать в рамках транзакции с полученным объектом класса-сущности из БД, а значит изменения будут автоматически зафиксированы)

TyutterinYakov reviewed (commented):
Добрый вечер! Отметил оставшиеся недочеты :) Хорошего вечера! Если возникнут вопросы - пиши в слак, буду рад помочь🌃

svlgerasimov committed:
refactor: убрать из BookingDtoIn NotNull, т.к. ограничение уже есть в BookingStartBeforeEnd

svlgerasimov committed:
fix: начало lastBooking может быть равно текущему времени

svlgerasimov committed:
refactor: убрать save при обновлении сущностей в транзакционных методах

TyutterinYakov reviewed (approved):
Добрый день! Поздравляю с успешной сдачей спринта! Сил и терпения на этом сложном, но интересном пути :) Успехов в обучении! "Жизнь – это большой холст и вы должны бросить на него всю краску какую только можете." Прекрасного дня!🌁
